### DISCUSSION TITLE
How to Implement Singleton Drawer with Different Sidebars Across Multiple Pages

### DISCUSSION THREAD

**Discussion by Typogram** [skeletonlabs/skeleton#2155] [General]
Hello everyone,

I'm having trouble understanding and implementing the drawer as a singleton in a responsive design. In my project, I have multiple pages each requiring different sidebars (drawers). For larger screens, a visible sidebar is present on the left, but on smaller screens, it’s hidden and replaced by a hamburger menu, which when clicked, opens the drawer from the left.

I comprehend that the drawer, among other components, is designed as a singleton. However, it's unclear to me why this is necessary and how to correctly use it in my case. I would like to learn and get guidance on how to effectively implement the singleton drawer to cater to varying menu items required by different pages in a mobile-responsive manner.

Any help, advice, or resources to guide me through this would be immensely appreciated. Thank you!

**Comment by endigo9740** [comment]
Hi @Typogram

Sure I can help with some of this.

We provide a tutorial which should cover the means to create a responsible sidebar, which appears fixed on large screens and appears in drawer a mobile. It's the same technique we use on the Skeleton documentation site in fact:
https://www.skeleton.dev/blog/how-to-implement-a-responsive-sidebar-drawer

> However, it's unclear to me why this is necessary

I developed this approach to help prevent redundancy. This is based on my experience creating web apps over the past 20+ years. In the wide majority of cases a single drawer with dynamic or "exchangeable" content cuts down on a lot of unnecessary repetition. It was also a design choice. There's rarely a silver bullet when you're building for an audience of the scale of Skeleton.

> I would like to learn and get guidance on how to effectively implement the singleton drawer to cater to varying menu items required by different pages in a mobile-responsive manner.

First, I'd encourage you to review this simple approach for swapping out the content within the drawer. If you only have a handful of pages, you might be able to abstract the sidebar content for each page to their own dedicated component. Then monitor the page route to determine which should be shown at a given time:
https://www.skeleton.dev/utilities/drawers#handling-contents

If you have lots of pages, I'd like to recommend a more advanced option. In addition to the tutorial I linked above, which cover the the basic concept of switching between fixed side + drawer, I'd encourage you to review how we handle our dynamic contents for our doc site drawer. In our case, the sidebar navigation isn't truly unique for every section, it's a standardized list, but with different list items. I'll link the important references below:

Implementing our abstracted Drawer and Sidebar components within an App Shell:
https://github.com/skeletonlabs/skeleton/blob/dev/sites/skeleton.dev/src/routes/%2Blayout.svelte#L214
https://github.com/skeletonlabs/skeleton/blob/dev/sites/skeleton.dev/src/routes/%2Blayout.svelte#L219

Our local drawer and sidebar components:
https://github.com/skeletonlabs/skeleton/blob/dev/sites/skeleton.dev/src/lib/components/DocsDrawer/DocsDrawer.svelte
https://github.com/skeletonlabs/skeleton/blob/dev/sites/skeleton.dev/src/lib/components/DocsSidebar/DocsSidebar.svelte

In our case, we listen to the page route via the SvelteKit `page` store, then select between a category of navigation links:
https://github.com/skeletonlabs/skeleton/blob/dev/sites/skeleton.dev/src/lib/components/DocsSidebar/DocsSidebar.svelte#L19

The category of links is then maintained here:
https://github.com/skeletonlabs/skeleton/blob/dev/sites/skeleton.dev/src/lib/links.ts

This file is also multi-purpose, as it's used to generate our search results as well. Hence the keywords. In our case search lives within a custom component modal:
https://github.com/skeletonlabs/skeleton/blob/dev/sites/skeleton.dev/src/lib/modals/DocsSearch/DocsSearch.svelte

Hopefully that's enough to get you started. If I can clarify on any specifics let me know. I'd also encourage you to review whether our drawer system fits your mental model. You're are actively encourage to swap out parts of Skeleton that do not fit your needs. If you wish to roll your own drawer system or incorporate a third party solution, go for it. Whatever works best for you.

**Comment by Typogram** [comment]
Hi Chris,

Firstly, I want to express my gratitude for the comprehensive tutorial you provided. It laid a solid foundation and clarified the essentials of implementing the drawer using conditions. Your detailed explanation was instrumental in helping me grasp the main concepts.

Regarding the singleton approach for the drawer, I have a few reservations. The drawer has versatile uses beyond just site navigation, such as for various configuration panels. In scenarios where multiple drawers could exist within the same page or across a multi-page application, having a singleton drawer might make the implementation awkward. Relying excessively on `$page` or other local conditions and funneling everything into the root layout tends to make the codebase harder to maintain and understand, especially when drawer content interacts closely with local pages.

### Exploring New Ideas: CSS Trick for Sidebar Reappearance

I experimented with a CSS trick to make the sidebar reappear with a click of a menu button. Although it lacks the backdrop or transitions, it simplifies the process and minimizes code repetition. This approach also ensures that the sidebar remains local. re-implementing them in a singleton root-level drawer does not seem worth the additional complexity and effort.

Considering this, I wonder if there's room to encapsulate some drawer-like logic directly into a new app sidebar component (as part of the AppShell cluster)? This could provide a more modular and maintainable way to implement the drawer functionality, catering to a variety of use cases without being confined to a singleton structure.

Thank you once again for the valuable insights shared in the tutorial, and I’m looking forward to any thoughts you might have on this feedback.

**Reply by endigo9740** [reply]
Hey, we always welcome feedback and constructive criticism. I agree with your points raised, a better means to create local instances or more fine tuned control within the page context would be beneficial. Rather than doing away with our current implementation and the singleton pattern, I think what we lean into the approach we came up for Component Modals. Which operate in a similar manner. These already solve this problem in a very flexible manner:

1. You can create a "registry" of reusable modals (think drawer content in this case) using reference components
2. You can directly supply a single component PER trigger (one-off drawer contents). Which would be perfect for the use case you describe.

As for WHEN this might occur, it would be a very notable rewrite of the current feature, which means this will likely need to be introduced alongside a major release of Skeleton. Given it's a major breaking change. However, we are considering breaking up and modularizing parts of Skeleton. Several features that fall under the "Utilities" category in our docs may end up as their own standalone library packages, separate from the core package.

I believe there is a way to combine Skeleton drawers/toasts/modals into into a singular library package for this use case. Which would mean they could share a lot of the same core feature between all three "flavors" of modals, including this new component-driven approach.

The other benefit to splitting this off, is we could introduce this at any point since it happens in parallel with the core library. But again, a big idea, so unfortunately no solid time frame can be provided for this. It'll come when we get to it. I would encourage you to keep an eye out for this in the future.

For now, I'd recommend whatever approach you think works best for your app. Skeleton provides a set of tools for you, but it's up to you to figure out how to architect your app. The idea you proposed might work. Give it a shot!
